api/provider/aws/test_aws/service_list_test.go

```diff
@@ -0,0 +1,188 @@
+
+	mockAWS := awsc.NewMockClient(ctrl)
+	tagStore := tag.NewMemoryStore()
+	mockConfig := mock_config.NewMockAPIConfig(ctrl)
+
+	mockConfig.EXPECT().Instance().Return("test").AnyTimes()
+
+	tags := models.Tags{
+		{
+			EntityID:   "env_id1",
+			EntityType: "environment",
+			Key:        "name",
+			Value:      "env_name1",
+		},
+		{
+			EntityID:   "env_id2",
+			EntityType: "environment",
+			Key:        "name",
+			Value:      "env_name2",
+		},
+		{
+			EntityID:   "svc_id1",
+			EntityType: "service",
+			Key:        "name",
+			Value:      "svc_name1",
+		},
+		{
+			EntityID:   "svc_id1",
+			EntityType: "service",
+			Key:        "environment_id",
+			Value:      "env_id1",
+		},
+		{
+			EntityID:   "svc_id2",
+			EntityType: "service",
+			Key:        "name",
+			Value:      "svc_name2",
+		},
+		{
+			EntityID:   "svc_id2",
+			EntityType: "service",
+			Key:        "environment_id",
+			Value:      "env_id1",
+		},
+		{
+			EntityID:   "svc_id3",
+			EntityType: "service",
+			Key:        "name",
+			Value:      "svc_name3",
+		},
+		{
+			EntityID:   "svc_id3",
+			EntityType: "service",
+			Key:        "environment_id",
+			Value:      "env_id2",
+		},
+		{
+			EntityID:   "svc_id4",
+			EntityType: "service",
+			Key:        "name",
+			Value:      "svc_name4",
+		},
+		{
+			EntityID:   "svc_id4",
+			EntityType: "service",
+			Key:        "environment_id",
+			Value:      "env_id2",
+		},
+	}
+
+	for _, tag := range tags {
+		if err := tagStore.Insert(tag); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// listClusterNames() calls ECS.ListClustersPages()
+	clusterARNs := []*string{
+		aws.String("arn:aws:ecs:region:012345678910:cluster/l0-test-env_id1"),
+		aws.String("arn:aws:ecs:region:012345678910:cluster/l0-test-env_id2"),
+		aws.String("arn:aws:ecs:region:012345678910:cluster/l0-bad-env_id1"),
+		aws.String("arn:aws:ecs:region:012345678910:cluster/bad"),
+	}
+
+	listClustersPagesFN := func(input *ecs.ListClustersInput, fn func(output *ecs.ListClustersOutput, lastPage bool) bool) error {
+		output := &ecs.ListClustersOutput{}
+		output.SetClusterArns(clusterARNs)
+		fn(output, true)
+
+		return nil
+	}
+
+	mockAWS.ECS.EXPECT().
+		ListClustersPages(gomock.Any(), gomock.Any()).
+		Do(listClustersPagesFN).
+		Return(nil)
+
+	// listClusterServiceNames() calls ECS.ListServicesPages() once for each cluster
+	clusterServices := map[string][]*string{
+		"l0-test-env_id1": []*string{
+			aws.String("arn:aws:ecs:region:012345678910:service/l0-test-svc_id1"),
+			aws.String("arn:aws:ecs:region:012345678910:service/l0-test-svc_id2"),
+		},
+		"l0-test-env_id2": []*string{
+			aws.String("arn:aws:ecs:region:012345678910:service/l0-test-svc_id3"),
+			aws.String("arn:aws:ecs:region:012345678910:service/l0-test-svc_id4"),
+		},
+	}
+
+	generateListServicesPagesFN := func(input *ecs.ListServicesInput) func(input *ecs.ListServicesInput, fn func(output *ecs.ListServicesOutput, lastPage bool) bool) error {
+		clusterName := *input.Cluster
```

Comment:
can't you just pass in `clusterName string` instead of `input *ecs.ListServicesInput`?

Reply:
Yeah, that's a bit simpler. ✅